toolchain/sem_ir/typed_insts.h

```diff
@@ -1341,6 +1341,18 @@ struct RequireCompleteType {
   TypeInstId complete_type_inst_id;
 };
 
+// A `require` declaration, such as `require Self impls Z`.
+struct RequireDecl {
+  static constexpr auto Kind =
+      InstKind::RequireDecl.Define<Parse::RequireDeclId>(
+          {.ir_name = "require_decl", .is_lowered = false});
+
+  // Always `type`.
+  TypeId type_id;
```

Comment:
I see there's a little similar with interfaces and named constraints, but my instinct would usually be that the type should be separate (e.g. a singleton, or a class decl vs class type split). What made you choose this particular typing approach?

Reply:
Closing in favor of https://github.com/carbon-language/carbon-lang/pull/6318#discussion_r2491852239